docs/design_proposals/configurable-transformable-allowlist.md

```diff
@@ -0,0 +1,74 @@
+# Configurable transformableAllowList
+
+* Author(s): Ke Zhu (@shawnzhu)
+* Design Shepherd: Tejal Desai (tejal29)
+* Date: 2021-07-16
+* Status: Draft
+
+## Objectives
+
+Configurable transformableAllowList for transforming manifests.
+
+## Background
+
+Skaffold can only transform manifests from a non-extensible allowlist. When
+using any CRD out of this allowlist, skaffold can not transform it.
+
+Open issues concerning this problem:
+
+* Image not recognized in crd k8s manifest ([#4081](https://github.com/GoogleContainerTools/skaffold/issues/4081))
+
+There was comments out of the above issue to make this allowlist extensible.
+
+The goal of this document is to create an agreement on the configuration option
+and specification to extend `skaffold.kubernetes.manifest.transformableAllowList`.
+
+## Design
+
+Add configuration `deploy.config.transformableAllowList` in `skaffold.yaml`:
+
+Notice that any new configuration option will be appended to existing allowlist. 
+
+### Detailed discussion
+
+Option in `skaffold.yaml`
+
+```YAML
+deploy:
+  config:
+    transformableAllowList:
+    - Group: example.com
+      Kind: Application
```

Comment:
It would be nice if this could be more compact.  And we should also be able to specify versions here too as they may have dramatically different shapes.  In Kubernetes, these objects are referred to as [_resource types_](https://kubernetes.io/docs/reference/using-api/api-concepts/#resource-versions), and they have versions.  

Then I think we should allow some configurability to determine whether (and where) we can perform certain transforms.  Currently Skaffold:
  - *applies labels:* Skaffold looks for any field named `metadata` and add a `labels` block if missing. This action breaks CRDs, which is why we blocked this behind an allowlist.
  - *rewrite images:* Skaffold looks for any field named `image`

So I suggest something like the following for the simple cases (rewrite everything based on field names):
```
- type: pod # no group, implicitly all versions
- type: batch/Job # group, implicitly all versions
- type: extensions/v1beta1/Deployment  # specific version
```

And we might try something like adding `images` and `labels` fields that provide a list of JSON-path-like paths. Maybe with `*` support?
```
- type: openfaas.com/v1/Function
  images: [spec.image]
  labels: [spec.metadata.labels, spec.labels]    # https://www.openfaas.com/blog/manage-functions-with-kubectl/
- type: apps/v1/Deployment
  images: [spec.template.spec.initContainers.*.image, spec.template.spec.containers.*.image]
  labels: [spec.metadata.labels, spec.template.metadata.labels]
```

Nit: these fields should be lower-cased in keeping with the rest (k8s yaml uses lower case)

Reply:
I like such compact configuration and the explict specification for `images` and `labels`.

How about this:

1. adopt the compact configuration
2. implement applies labels so it only rewrite allowed labels of resource types in allowlist
3. only rewrite image of resource types in allowlist

Any guidance on the implementation like which module to change would also be appreicated!

Reply:
https://github.com/GoogleContainerTools/skaffold/pull/6236/commits/6a048ee83deff7d30d75edd54f2652147d980feb